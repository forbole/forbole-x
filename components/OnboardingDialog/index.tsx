import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  IconButton,
  Typography,
} from '@material-ui/core'
import useTranslation from 'next-translate/useTranslation'
import React from 'react'
import ClossIcon from '../../assets/images/icons/icon_cross.svg'
import useStyles from './styles'
import PasswordInput from '../PasswordInput'
import useIconProps from '../../misc/useIconProps'
import useIsMobile from '../../misc/useIsMobile'

interface OnboardingDialogProps {
  open: boolean
  onClose(): void
  onSubmit(password: string): void
}

const OnboardingDialog: React.FC<OnboardingDialogProps> = ({ open, onClose, onSubmit }) => {
  const { t } = useTranslation('common')
  const classes = useStyles()
  const iconProps = useIconProps()
  const isMobile = useIsMobile()
  const [password, setPassword] = React.useState('')
  const [confirmPassword, setConfirmPassword] = React.useState('')
  const [isConfirmingPassword, setIsConfirmingPassword] = React.useState(false)
  const [error, setError] = React.useState('')

  const onButtonClick = React.useCallback(() => {
    setError('')
    if (!isConfirmingPassword) {
      if (password.length < 6) {
        setError(t('invalid password'))
      } else {
        setIsConfirmingPassword(true)
      }
    } else if (password !== confirmPassword) {
      setError(t('invalid confirm password'))
    } else {
      onSubmit(password)
    }
  }, [isConfirmingPassword, password, confirmPassword])
  return (
    <Dialog fullWidth open={open} onClose={onClose} fullScreen={isMobile}>
      <IconButton className={classes.closeButton} onClick={onClose}>
        <ClossIcon {...iconProps} />
      </IconButton>
      <DialogTitle>
        {isConfirmingPassword ? t('confirm password title') : t('set password title')}
      </DialogTitle>
      <form
        noValidate
        onSubmit={(e) => {
          e.preventDefault()
          onButtonClick()
        }}
      >
        <DialogContent>
          <DialogContentText>
            {isConfirmingPassword
              ? t('confirm password description')
              : t('set password description')}
          </DialogContentText>
          <PasswordInput
            placeholder={t('password')}
            value={isConfirmingPassword ? confirmPassword : password}
            error={!!error}
            helperText={error}
            onChange={(e) =>
              (isConfirmingPassword ? setConfirmPassword : setPassword)(e.target.value)
            }
          />
          <Typography className={classes.passwordRequirement} variant="body2">
            {isConfirmingPassword ? t('confirm password caption') : t('password caption')}
          </Typography>
        </DialogContent>
        <DialogActions>
          <Button className={classes.button} variant="contained" color="primary" type="submit">
            {isConfirmingPassword ? t('confirm') : t('next')}
          </Button>
        </DialogActions>
      </form>
    </Dialog>
  )
}

export default OnboardingDialog
