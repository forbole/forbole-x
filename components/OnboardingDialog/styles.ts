import { makeStyles } from '@material-ui/core/styles';
import { CustomTheme } from '../../misc/theme';

const useStyles = makeStyles(
  (theme: CustomTheme) => ({
    passwordRequirement: {
      marginTop: theme.spacing(1),
    },
    button: {
      flex: 1,
      margin: theme.spacing(4, 2),
      marginTop: theme.spacing(20),
      color: 'white',
    },
    closeButton: {
      position: 'absolute',
      top: theme.spacing(2),
      right: theme.spacing(2),
    },
  }),
  {
    name: 'HookGlobalStyles',
    index: 2,
  },
);

export default useStyles;
