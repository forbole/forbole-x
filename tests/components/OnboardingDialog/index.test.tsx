import React from 'react';
import renderer from 'react-test-renderer';
import OnboardingDialog from '../../../components/OnboardingDialog';

const onClose = jest.fn();
const onSubmit = jest.fn();

jest.mock('../../../components/PasswordInput', () => 'PasswordInput');
jest.mock('@material-ui/core/Dialog', () => props => <div id="dialog" {...props} />);

describe('component: OnboardingDialog', () => {
  it('renders open state correctly', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders closed state correctly', () => {
    const component = renderer.create(
      <OnboardingDialog open={false} onClose={onClose} onSubmit={onSubmit} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('calls onClose when IconButton is clicked', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    renderer.act(() => {
      component.root
        .findByProps({
          className: 'MuiIconButton-label',
        })
        .parent.props.onClick();
    });
    expect(onClose).toBeCalled();
  });
  it('renders error state when submit invalid password', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders confirming password state when submit valid password', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    renderer.act(() => {
      component.root.findByType('PasswordInput' as any).props.onChange({
        target: {
          value: '123456',
        },
      });
    });
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders error state when submit invalid confirm password', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    renderer.act(() => {
      component.root.findByType('PasswordInput' as any).props.onChange({
        target: {
          value: '123456',
        },
      });
    });
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    renderer.act(() => {
      component.root.findByType('PasswordInput' as any).props.onChange({
        target: {
          value: '123455',
        },
      });
    });
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly when submit valid confirm password', () => {
    const component = renderer.create(
      <OnboardingDialog open onClose={onClose} onSubmit={onSubmit} />,
    );
    renderer.act(() => {
      component.root.findByType('PasswordInput' as any).props.onChange({
        target: {
          value: '123456',
        },
      });
    });
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    renderer.act(() => {
      component.root.findByType('PasswordInput' as any).props.onChange({
        target: {
          value: '123456',
        },
      });
    });
    renderer.act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: jest.fn() });
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

afterEach(() => {
  jest.clearAllMocks();
});
