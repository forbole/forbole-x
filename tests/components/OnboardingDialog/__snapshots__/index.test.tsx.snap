// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: OnboardingDialog renders closed state correctly 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={false}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:set password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:set password description
      </p>
      <PasswordInput
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value=""
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:next
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`component: OnboardingDialog renders confirming password state when submit valid password 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:confirm password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:confirm password description
      </p>
      <PasswordInput
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value=""
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:confirm password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:confirm
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`component: OnboardingDialog renders correctly when submit valid confirm password 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:confirm password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:confirm password description
      </p>
      <PasswordInput
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value="123456"
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:confirm password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:confirm
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`component: OnboardingDialog renders error state when submit invalid confirm password 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:confirm password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:confirm password description
      </p>
      <PasswordInput
        error={true}
        helperText="common:invalid confirm password"
        onChange={[Function]}
        placeholder="common:password"
        value="123455"
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:confirm password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:confirm
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`component: OnboardingDialog renders error state when submit invalid password 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:set password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:set password description
      </p>
      <PasswordInput
        error={true}
        helperText="common:invalid password"
        onChange={[Function]}
        placeholder="common:password"
        value=""
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:next
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`component: OnboardingDialog renders open state correctly 1`] = `
<div
  fullScreen={false}
  fullWidth={true}
  id="dialog"
  onClose={[MockFunction]}
  open={true}
>
  <button
    className="MuiButtonBase-root MuiIconButton-root HookGlobalStyles-closeButton-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        data-file-name="SvgIcon_cross"
        fill="#e0e0e0"
        height={16}
        width={16}
      />
    </span>
  </button>
  <div
    className="MuiDialogTitle-root"
  >
    <h2
      className="MuiTypography-root MuiTypography-h6"
    >
      common:set password title
    </h2>
  </div>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <p
        className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        common:set password description
      </p>
      <PasswordInput
        error={false}
        helperText=""
        onChange={[Function]}
        placeholder="common:password"
        value=""
      />
      <p
        className="MuiTypography-root HookGlobalStyles-passwordRequirement-1 MuiTypography-body2"
      >
        common:password caption
      </p>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained HookGlobalStyles-button-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          common:next
        </span>
      </button>
    </div>
  </form>
</div>
`;
